qiskit/utils/deprecation.py

```diff
@@ -77,3 +92,73 @@ def _rename_kwargs(func_name, kwargs, kwarg_map, category: Type[Warning] = Depre
                 )
 
                 kwargs[new_arg] = kwargs.pop(old_arg)
+
+
+def _extend_docstring(func, version, kwarg_map):
+    """kwarg_map[None] means message in no kwarg, and it will be
+    append to the long desscription"""
+    docstr = func.__doc__
+    if docstr:
+        docstr_lines = docstr.expandtabs().splitlines()
+    else:
+        docstr_lines = ["DEPRECATED"]
+
+    indent = 1000
+    first_empty_line = None
+    for line_no, line in enumerate(docstr_lines[1:], start=1):
+        stripped = line.lstrip()
+        if stripped:
+            indent = min(indent, len(line) - len(stripped))
+        else:
+            if first_empty_line is None:
+                first_empty_line = line_no
+    if first_empty_line is None:
+        first_empty_line = len(docstr_lines)
+    spaces = ""
+    if indent != 1000:
+        spaces = " " * indent
+
+    new_doc_str_lines = docstr_lines[:first_empty_line]
+    if None in kwarg_map:
+        new_doc_str_lines += ["", spaces + f".. deprecated:: {version}"]
+        for msg_line in kwarg_map[None]:
+            new_doc_str_lines.append(spaces + "  " + msg_line)
+
+    arg_indent = indent
+    args_section = False
+    deprecated_arg = False
+    for docstr_line in docstr_lines[first_empty_line:]:
+        stripped = docstr_line.lstrip()
+        current_indent = len(docstr_line) - len(stripped)
+        if args_section:
```

Comment:
You might consider simplifying the deep nesting and inter-loop-iteration communication by using co-routines with a shared iterator object.  Here's a rough example with simplified internals that just prints a prefix when it's in the `args` section:
```python
def handle_arg_section(iterator):
    spaces, n = None, 0
    for line in iterator:
        if spaces is None:
            n = len(line) - len(line.lstrip())
            spaces = " " * n
        if line[:n].strip():
            return line
        yield f"{spaces}[in args]: {line[n:]}"
    return None

def handle_docstring(string):
    iterator = iter(string.splitlines())
    for line in iterator:
        stripped = line.strip()
        if stripped == "Args:":
            yield stripped
            next_line = yield from handle_arg_section(iterator)
            if next_line is not None:
                yield next_line
        else:
            yield stripped
```

Now

```python
print("\n".join(_handle_docstring("""
hello, world
this is my string

Args:
    arg1: pass
    arg2: what else is going on here?
       this argument wraps around, and
       even has multiple paragraphs.

       Here's another paragraph.

back out in the main docstring""")))
```
produces
```text
hello, world
this is my string

Args:
    [in args]: arg1: pass
    [in args]: arg2: what else is going on here?
    [in args]:    this argument wraps around, and
    [in args]:    even has multiple paragraphs.
    [in args]:
    [in args]:    Here's another paragraph.
    [in args]:
back out in the main docstring
```

(That also handles blank lines and continuation lines in the parameter definitions better.)

It's not _necessarily_ shorted (though it often will be), but it's much easier to read and maintain the separation of concerns of the two components.  You can even extend this to add a special handler for individual arguments in the args block as well, if you like, to keep the separation greater.

(This is similar in spirit to a recursive-descent parser.)